qcodes/instrument_drivers/Harvard/Decadac.py

```diff
@@ -56,23 +55,67 @@ def __init__(self, name, port, slot, timeout=2, baudrate=9600,
-        self.ramp_time = 100
+        self._ramp_state = False
+        self._ramp_time = 100
+        self._voltranges = [1, 1, 1, 1]
+        self._offsets = [0, 0, 0, 0]
 
-        # initialise hardware settings
-        self.voltranges = [1, 1, 1, 1]
+        self.add_parameter('mode',
+                           label='Output mode',
+                           set_cmd='B {}; M {};'.format(self.slot, '{}'),
+                           vals=vals.Enum(0, 1))
 
-        # set up four channels as qcodes parameters
+        # channels
         for channelno in range(4):
-
-            self.add_parameter('volt{:d}'.format(channelno),
+            self.add_parameter('ch{}_voltage'.format(channelno),
                                get_cmd=partial(self._getvoltage,
                                                channel=channelno),
                                set_cmd=partial(self._setvoltage,
                                                channel=channelno),
                                label='Voltage',
                                units='V')
 
+            self.add_parameter('ch{}_voltrange'.format(channelno),
+                               get_cmd=partial(self._getvoltrange, channelno),
+                               set_cmd=partial(self._setvoltrange, channelno),
+                               vals=vals.Enum(1, 2, 3))
+
+            self.add_parameter('ch{}_offset'.format(channelno),
+                               get_cmd=partial(self._getoffset, channelno),
+                               set_cmd=partial(self._setoffset, channelno),
+                               label='Channel {} offset'.format(channelno),
+                               units='V',
+                               docstring="""
+                                         The offset is applied to the channel.
+                                         E.g. if ch1_offset = 1 and ch_voltage
+                                         is set to 1, the instrument is told to
+                                         output 2 volts.
+                                         """)
+
+        self.add_parameter('mode',
+                           label='Output mode',
+                           set_cmd='B {}; M {};'.format(self.slot, '{}'),
+                           vals=vals.Enum(0, 1),
+                           docstring="""
+                                     The operational mode of the slot.
+                                     0: output off, 1: output on.
+                                     """)
+
+        # initialise hardware settings
+        self.mode.set(1)
+
+    def _getoffset(self, n):
+        return self._offsets[n]
+
+    def _setoffset(self, n, val):
+        self._offsets[n] = val
+
+    def _getvoltrange(self, n):
+        return self._voltranges[n]
+
+    def _setvoltrange(self, n, val):
+        self._voltranges[n] = val
+
     def _getvoltage(self, channel):
         """
         Function to query the voltage. Flushes the message queue in that
```

Comment:
line 130 threw an error if I tried to ask for voltage before setting it the first time, also really crazy behaviour sometimes with this call where it just straight up lies about what voltage it's outputting and then tells the truth if you ask a second time...perhaps part of the same problem with something message-queue-like